### DISCUSSION TITLE
Question about `Add support for diff-patching filter lists`

### DISCUSSION THREAD

**Discussion by Nomes77** [uBlockOrigin/uAssets#20386] [General] [answered]
When [`Add support for diff-patching filter lists` ](https://github.com/gorhill/uBlock/commit/d05ff8ffebf8a35a9048c52837fa60a7a55f222f) is good and well implemented in the main releases, do you then welcome it if filterlist maintainers also implement it. Or is it not necessary?

Context:
- KOR-1: [![](https://data.jsdelivr.com/v1/package/gh/List-KR/List-KR/badge)](https://www.jsdelivr.com/package/gh/List-KR/List-KR) | 447GB/month
- NLD-0: [![](https://data.jsdelivr.com/v1/package/gh/EasyDutch-uBO/EasyDutch/badge)](https://www.jsdelivr.com/package/gh/EasyDutch-uBO/EasyDutch) | 80GB/month
- NOR-0: [![](https://data.jsdelivr.com/v1/package/gh/DandelionSprout/adfilt/badge)](https://www.jsdelivr.com/package/gh/DandelionSprout/adfilt) | 81GB/month
- RUS-0: [![](https://data.jsdelivr.com/v1/package/gh/easylist/ruadlist/badge)](https://www.jsdelivr.com/package/gh/easylist/ruadlist) | 4.187TB/month

**Comment by gorhill** [comment]
> do you then welcome it if filterlist maintainers also implement it

Sure, if the filter list contains all that is needed according to spec, uBO will try to patch the list.

**Reply by Nomes77** [reply]
That is a quick reaction. 
As you can see in our [EasyDutchCDN](https://github.com/EasyDutch-uBO/EasyDutchCDN) we applied it.
May I propose then the following: 

'0' is the `contentURL`, which is just the normal list
'1-3' are `cdnURLs`, which are all minified
 
0. https://easydutch-ubo.github.io/EasyDutch/EasyDutch.txt
1. https://easydutchcdn.pages.dev/EasyDutch.all.txt
2. https://cdn.jsdelivr.net/gh/EasyDutch-uBO/EasyDutchCDN@main/EasyDutch.all.txt
3. https://cdn.statically.io/gh/EasyDutch-uBO/EasyDutchCDN/main/EasyDutch.all.txt

#### (See: "Note on Data usage" in https://github.com/orgs/EasyDutch-uBO/discussions/44)

**Reply by krystian3w** [reply]
Maybe someone made script in bash/python for community - compatible with mechanisms like GitHub Actions.

**Reply by gorhill** [reply, marked as answer]
I've updated `assets.dev.json` with your new URLs, will see if this works. I am still addressing small issues. So for now only the dev build will try to fetch these new URLs, I will update the ones in `assets.json` once the new updater reaches stable.

**Comment by dportvine** [comment]
How to do it with **RU AdList** ?

**Reply by Nomes77** [reply]
To use [EasyDutch](https://github.com/EasyDutch-uBO/EasyDutch) as an example.

1. make the file `tools/RuAdList-uBO.template`:
```
! Title: RU AdList for uBO only
! Last modified: %timestamp%
! Expires: 2 days (фильтр обновляется автоматически каждые 2 дня)
! Diff-Path: %diffpath%#ruadlistubo
! Diff-Name: ruadlistubo
! Diff-Expires: 6 hours
! Homepage: https://forums.lanik.us/viewforum.php?f=102
!
! Комбинированная версия списков фильтров RU AdList против рекламы
! и экспериментальных списков RU AdList CSS Fixes и RU AdList JS Fixes
! На данный момент нормально работает только в uBlock Origin
!

%include ruadlist:advblock/adservers.txt%

%include ruadlist:advblock/first_level.txt%

%include ruadlist:advblock/general_block.txt%

%include ruadlist:advblock/general_hide.txt%

%include ruadlist:advblock/popup.txt%

%include ruadlist:advblock/specific_antisocial.txt%

%include ruadlist:advblock/specific_block.txt%

%include ruadlist:advblock/specific_hide.txt%

%include ruadlist:advblock/specific_special.txt%

%include ruadlist:advblock/thirdparty.txt%

%include ruadlist:advblock/whitelist.txt%

%include ruadlist:css-fixes-experimental.txt%

%include ruadlist:js-fixes-experimental.txt%

%include ruadlist:AWRL-non-sync.txt%

```
2. Add [`make-diffpatch.sh`](https://github.com/EasyDutch-uBO/EasyDutch/blob/main/tools/make-diffpatch.sh) to the same folder
3. Add `make-list.js` to the same folder (please copy the file from [uAssets](https://github.com/uBlockOrigin/uAssets/blob/master/tools/easylist/make-easylist.js), because I chanced some things for our specific (EasyDutch) needs.).
4. Add [`update-diffpatches.sh`](https://github.com/EasyDutch-uBO/EasyDutch/blob/main/tools/update-diffpatches.sh) and change line 36 from `git clone -q --single-branch --branch $PREVIOUS_VERSION --depth=1 https://github.com/EasyDutch-uBO/EasyDutch.git "$OLD_REPO" 2>/dev/null` to `git clone -q --single-branch --branch $PREVIOUS_VERSION --depth=1 https://github.com/easylist/ruadlist.git "$OLD_REPO" 2>/dev/null`
5. Make a `gh-pages` branch in which you add a empty file called: `RuAdList-uBO.min.txt`
Also add a empy file called `version` and make the folder `patches` by adding `patches/README.md`.
7. In the `master` branch add the following under `.github/workflows/main.yml`:
```
name: Deploy to gh-pages branch

on:
  push:
    branches:
      - master
    paths:
      - 'RuAdList-uBO.txt'
      - 'advblock/**'
      - 'css-fixes-experimental.txt'
      - 'js-fixes-experimental.txt'
      - 'AWRL-non-sync.txt'
      - 'tools/RuAdList-uBO.template'

# https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-only-cancel-in-progress-jobs-or-runs-for-the-current-workflow
concurrency: 
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

permissions:
  contents: read

jobs:
  publish:
    permissions:
      contents: write  # for Git to git push
    name: Publish lists
    runs-on: ubuntu-latest
    steps:
      - name: Clone RuAdList
        uses: actions/checkout@v3
        with:
          ref: gh-pages
      - name: Generate version string
        run: |
          TAGNAME=$(date -u "+%Y.%-m.%-d.")$(date -u "+%H*60+%M" | bc)
          echo "TAGNAME=$TAGNAME" >> $GITHUB_ENV
          echo "Version string is $TAGNAME"
      - name: Copy shell script from EasyList/RuAdList
        run: |
          TMPDIR=$(mktemp -d)
          git clone --depth=1 --single-branch --branch=master https://github.com/easylist/ruadlist.git $TMPDIR
          cp -v $TMPDIR/tools/make-diffpatch.sh .
          cp -v $TMPDIR/tools/update-diffpatches.sh .
      - name: Copy filter lists to gh-pages
        run: |
          TMPFILE=$(mktemp -d)
          git clone --depth=1 https://github.com/easylist/ruadlist.git $TMPFILE
          pushd $TMPFILE > /dev/null
          echo "*** RuAdList: Assembling RuAdList.txt"
          node ./tools/make-list.js in=./tools/RuAdList-uBO.template out=./RuAdList-uBO.min.txt minify=1
          popd > /dev/null
          cp $TMPFILE/RuAdList-uBO.min.txt RuAdList-uBO.min.txt
      - name: Patch last-updated field
        run: |
          DATE=$(date -Ru)
          for f in $(git diff --name-only); do
            STAT=$(git diff --numstat $f | sed -r '/^1\s+1\s+/d')
            if [[ -n $STAT ]]; then
              sed -ir "0,/^! Last updated: /s/^\(! Last updated: \)%timestamp%/\\1$DATE/" $f
            else
              git checkout -q $f
            fi
          done
      - name: Create new patch
        run: |
          chmod 755 ./make-diffpatch.sh
          ./make-diffpatch.sh "${{ env.TAGNAME }}"
      - name: Update existing patch
        run: |
          chmod 755 ./update-diffpatches.sh
          ./update-diffpatches.sh "${{ env.TAGNAME }}"
      - name: Commit changes (if any)
        run: |
          if [[ -z $(git diff --name-only --cached) ]]; then exit 0; fi
          git config user.name "GitHub-Actions"
          git config user.email "<>"
          git commit -m "Update modified filter lists to ${{ env.TAGNAME }}"
          git push origin gh-pages
          git tag ${{ env.TAGNAME }}
          git push origin ${{ env.TAGNAME }}
```

**Reply by gorhill** [reply]
Note that the branch does not have to be called `gh-pages`, it can be called whatever you want. I had `gh-pages` on my side because historically I had enabled  "GitHub Pages" in repo settings, and that's the default name and I used that branch for distributing lists, but you do not need "GitHub Pages". just a different branch. ~~It doesn't even have to be another branch, it can be a folder under the main repo.~~ not a good idea in retrospect, a separate branch is better, but can be named whatever.

**Reply by krystian3w** [reply]
How to use this on only master branch?

**Reply by Nomes77** [reply]
I think just push to the `master` branch, so `git push origin master` in stead of `git push origin gh-pages`.
And of course put every file/folder I mentioned in point 5 in the `master` branch.

**Reply by gorhill** [reply]
Probably a disadvantage of using the master branch is that you always have to mind remote changes to the repo before pushing yours. Having another branch keeps all the remote changes confined to that branch, and also won't pollute the git history of the master branch, the git history of the content of the branch receiving the diff-able version of the list is not really of any interest, it could be wholly deleted and recreated in the future.

**Comment by gorhill** [comment]
I updated the bash scripts to encode the `! Diff-Name:` value directly into the `! Diff_Path:` value, so that there is no need for `! Diff-Name:`. I feel this is a better approach than having to keep track of a separate field, and the name of a specific diff-block inside the patch fits well the semantic of the hash portion of a URL.

Modification to bash scripts and examples of the new way to encode diff-block name: <https://github.com/uBlockOrigin/uAssets/commit/f34a1039f959fa5cefb869982c6f50fb5588d17e>

There need to be a transition, so for a while uBO itself will still use `! Diff-Name:` if no hash portion can be found in `! Diff-Path:`.

**Reply by dimisa-RUAdList** [reply]
Thanks to @dportvine for directing me to this discussion. And thanks to @Nomes77 for the instructions. I have little experience working with GitHub; previously, for many years I worked at `hg.adblockplus.org`. Do I need to change anything in the described procedure given the changes made in [this commit](https://github.com/uBlockOrigin/uAssets/commit/f34a1039f959fa5cefb869982c6f50fb5588d17e)?

**Reply by Nomes77** [reply]
@dimisa-RUAdList No/Yes. I changed the `tools/RuAdList-uBO.template` as described in step 1 to fit the given commit.

**Reply by JohnyP36** [reply]
@dimisa-RUAdList 
> Do I need to change anything in the described procedure given the changes made in [this commit](https://github.com/uBlockOrigin/uAssets/commit/f34a1039f959fa5cefb869982c6f50fb5588d17e)?

The step 1 given above has changed after gorhill's commit.

Previous code
```
! Diff-Path: %diffpath%
! Diff-Name: ruadlistubo
! Diff-Expires: 6 hours
```

Now
```
! Diff-Path: %diffpath%#ruadlistubo
! Diff-Name: ruadlistubo
! Diff-Expires: 6 hours
```

Also make sure to copy the changes that are made in the [`tools/make-diffpatch.sh` file](https://github.com/uBlockOrigin/uAssets/commit/f34a1039f959fa5cefb869982c6f50fb5588d17e#diff-d19fb39935db9362530400c9b91713ccf9a6aff8aef92195644e9b12b9a8af25) and the [`tools/update-diffpatches.sh` file](https://github.com/uBlockOrigin/uAssets/commit/f34a1039f959fa5cefb869982c6f50fb5588d17e#diff-549e1455f32cb6ea8cf4b3fd505d2f108e93a41c0296d116b428be3179847679).

**Reply by gorhill** [reply]
`! Diff-Name:` field can be removed once `#ruadlistubo` is added and v1.53.1.4 is out for all ([chromium still is at v1.53.1b1](https://chrome.google.com/webstore/detail/ublock-origin-development/cgbcahbpdhpcegmbfconppldiemgcoii)).

**Reply by dimisa-RUAdList** [reply]
Ok, I think I've done everything. How to check the result?

**Reply by gorhill** [reply]
Your list is empty: <https://github.com/easylist/ruadlist/blob/diff-pages/RuAdList-uBO.min.txt>

Looks like you created a branch just for the sake of hosting the diff-updatable list, but then you have also everything else in that branch. Probably best to create a repo which purpose is just to serve the diff-updatable list, i.e. something like `ruadlistCDN`, or maybe even `EasylistCDN` if the plan is to also have all of EasyList lists become diff-updatable at the source.

**Reply by dimisa-RUAdList** [reply]
As I wrote earlier, I have no experience with Github. I just followed step 5 of the [instructions](https://github.com/uBlockOrigin/uAssets/discussions/20386#discussioncomment-7453145):
> Make a gh-pages branch in which you add a empty file called: RuAdList-uBO.min.txt

If you need to add or correct something, I'm ready to do it.

**Reply by gorhill** [reply]
It's difficult to do by proxy, when I setup things it does happen I have to go through trials and errors. How about you create a repo ruadlistCDN, give me commit right and from there I will try to setup things myself. ONce all setup, you look at the commit to see how this was done.

**Reply by dimisa-RUAdList** [reply]
No problem. Please describe the steps that I need to complete.

**Reply by gorhill** [reply]
If you are not allowed to create a repo at easylist organisation, you will have to ask @ryanbr or whoever has admin rights to create one, say `ruadlistCDN`, then to give me and you commit rights into it.

**Reply by dimisa-RUAdList** [reply]
Ok, while we are waiting for @ryanbr response, does it make sense to delete all the files and folders created according to the instructions (including the `diff-pages` branch)?

**Reply by gorhill** [reply]
I think it's best to just have a dedicated repo instead of a branch, for uAssets, the diff-updatable lists are all at uAssetsCDN.

**Reply by dimisa-RUAdList** [reply]
In general, I rolled back for now.

**Reply by krystian3w** [reply]
And in one CDN repo can be shared many lists?

Examples:

- Polish anti frauds and scam list (e.g. rejected in URL Hause projects)
- Polish Annoyances list for anti-fans time to address sites/bugs in Fanboy Annoyances
- Polish RSS/Google news list due nobody splited Fanboy Social by move RSS to diffrent file

**Reply by gorhill** [reply]
@dimisa-RUAdList, if you want we can create the CDN repo in your account, doesn't look like its going to happen at easylist.

**Reply by dimisa-RUAdList** [reply]
@gorhill Yes I agree. But I need step by step instructions.

**Reply by gorhill** [reply]
Create a new repo in your account, named `RUAdListCDN`.

![image](https://github.com/uBlockOrigin/uAssets/assets/585534/fc9d2c09-b63a-4bc2-baff-eb1522a3761b)

Fill in description, agree to "Add a README file", pick a license.

**Reply by dimisa-RUAdList** [reply]
Public or Private?
<details>
<summary>Screenshot(s)</summary>

![Repo new](https://github.com/uBlockOrigin/uAssets/assets/20126984/98891818-2a12-45bc-bbca-0b976b296f3e)

</details>

And which license should I choose?

`.gitignore template` what should the value be?

**Reply by gorhill** [reply]
Public. For uAssets we picked GPLv3. Picking the license at that point will automatically generate the file in the repo.

**Reply by dimisa-RUAdList** [reply]
`.gitignore template` - None?

**Reply by gorhill** [reply]
For uAssetsCDN, we don't have a `.gitignore`.

**Reply by dimisa-RUAdList** [reply]
Ok, done: https://github.com/dimisa-RUAdList/RUAdListCDN

What to do next?

**Reply by gorhill** [reply]
I will submit a pull request, this might take a while though, so probably tomorrow.

**Reply by krystian3w** [reply]
It can't be skipped license file due protect wide spectrum blockers intellectual property laws?

**Comment by JohnyP36** [comment]
In the [`uAssets   tools/easylist/make-easylist.js`](https://github.com/uBlockOrigin/uAssets/blob/master/tools/easylist/make-easylist.js) is this code
```js
    // Remove sublist header information: the importing list will provide its
    // own header.
    text = text.trim().replace(/^(?:!\s+[^\r\n]+?(?:\r\n|\n))+/s, '');
```

Which I also like to use to minify my filterlist. 
The problem is that my sublist header info is 
```
[uBlock Origin]
! Last modified: %timestamp%
```
I know I have to change the regex, but somehow I don't get it working. 

Can some help me with this? 

---
#### I'm sorry if this discussion is the wrong place to ask this question.

**Reply by Nomes77** [reply]
`^(?:(?:\[|!\s+)[^\r\n]+?(?:\r\n|\n))+`

**Reply by JohnyP36** [reply]
Thanks. I made this `/^(?:(?:\[|!\s+[^0-9])[^\r\n]+?(?:\r\n|\n)){2}/s` *(used [regex101](https://regex101.com/))*

**Comment by ghost** [comment]
I would suggest running these scripts through shellcheck, they may fail on spaces.

**Reply by krystian3w** [reply]
You means validator/compilator or fix comments by Grammarly (removal of the local English dialect or obvious typos)?

**Reply by ghost** [reply]
Bash variable quoting.

**Reply by gorhill** [reply]
[Cool stuff](https://www.shellcheck.net/). I am not a bash shell script guru, so I am just fumbling my way around with search engine without knowing whether what I am doing is the correct way to do it. This will be helpful.
